go/ai/generate.go

```diff
@@ -86,8 +86,183 @@ func LookupModel(provider, name string) *Model {
 	return (*Model)(core.LookupActionFor[*GenerateRequest, *GenerateResponse, *GenerateResponseChunk](atype.Model, provider, name))
 }
 
-// Generate applies the [Model] to some input, handling tool requests.
-func (m *Model) Generate(ctx context.Context, req *GenerateRequest, cb ModelStreamingCallback) (*GenerateResponse, error) {
+// GenerateParams represents various params of the Generate call.
+type GenerateParams struct {
+	Request      *GenerateRequest
+	Stream       ModelStreamingCallback
+	History      []*Message
+	SystemPrompt *Message
+}
+
+// GenerateOption configures params of the Generate call.
+type GenerateOption func(req *GenerateParams) error
```

Comment:
Great. 
Next question: Do you want this to be exported?

If yes, then anyone can write their own GenerateOption and use it. But we have one more exported symbol (actually several, because the fields are exported.) A bigger API surface means more for people to understand. Even though 99% (or maybe 100%) of people won't care about this type and its fields and will never use it directly.

If no, then you control the options, no one can write their own, and no one sees this type (except as an arg to the GenerateOption function, which they will happily ignore).

I would always pick no: make this unexported. You can always change your mind and export it later (although it would technically be breaking to change the signature of the GenerateOption function, it's extremely unlikely that anyone will be affected).

Reply:
agreed. unexported.